tools/fs/FS_Dump_Parser/src/smartfs_dump_visualizer/controllers/SmartFSOrganizer.java

```diff
@@ -0,0 +1,107 @@
+package smartfs_dump_visualizer.controllers;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import smartfs_dump_parser.data_model.EntryType;
+import smartfs_dump_parser.data_model.ModelProvider;
+import smartfs_dump_parser.data_model.Sector;
+import smartfs_dump_parser.data_model.SectorStatus;
+import smartfs_dump_parser.data_model.SmartFile;
+
+public class SmartFSOrganizer {
+
+	public static void createDirectoryHierarchy() {
+		List<SmartFile> queue = new ArrayList<SmartFile>();
+		SmartFile root = ModelProvider.INSTANCE.getRootDir();
+		if (root == null) {
+			System.out.println("Root is null");
+		}
+		ModelProvider.getTopDummyDirectory().getEntries().add(root);
+		queue.add(root);
+
+		while (!queue.isEmpty()) {
+			SmartFile sf = queue.remove(0);
+			if (analyzeChildrenEntries(sf) > 0) {
+				queue.addAll(sf.getEntries());
+			}
+		}
+		System.out.println("Directory hierarchy is organized.");
+	}
+
+	private static int analyzeChildrenEntries(SmartFile sf) {
+		if (sf.getSectorList().size() == 0) {
+			System.out.println("Warning: No sectors are specified for this entry: " + sf.getFileName());
+			return -1;
+		}
+
+		Sector currSector = sf.getSectorList().get(0);
+		byte[] flashData = currSector.getFlashData();
+		int offset = ModelProvider.getSectorHeaderSize();
+
+		// Get the information of 'struct smartfs_chain_header_s'
+		int type = ModelProvider.makePositiveValue(flashData[offset]);
+		if (type == 0x0) {
+			System.out.println("File " + sf.getFileName() + " has no children..");
+			return sf.getEntries().size();
+		}
+
+		// Get the information of 'struct smartfs_entry_header_s'
+		offset += ModelProvider.getChainHeaderSize();
+		while ((offset + ModelProvider.getEntryHeaderSize() < ModelProvider.getSectorSize())
+				&& (ModelProvider.makePositiveValue(flashData[offset]) != 255)) {
+			boolean isDirectory = true;
+			if ((flashData[offset] & 0x20) == 0x0) {
+				isDirectory = false;
+			}
+			offset += ModelProvider.getEntryHeaderSize();
+
+			char[] fileNameArray = new char[ModelProvider.getFileNameLength()];
+			for (int i = 0; i < ModelProvider.getFileNameLength(); i++) {
+				int character = ModelProvider.makePositiveValue(flashData[offset + i]);
+				if (character > 0) {
+					fileNameArray[i] = (char) ('a' + character - 97);
+				} else {
+					fileNameArray[i] = '\0';
+				}
+			}
+			// TODO: Fill sectorList appropriately
+			SmartFile child = new SmartFile(new String(fileNameArray).trim(),
+					(isDirectory ? EntryType.DIRECTORY : EntryType.FILE), sf, new ArrayList<Sector>());
+			sf.getEntries().add(child);
+			offset += +ModelProvider.getFileNameLength();
+		}
+		return sf.getEntries().size();
+	}
+
+	public static List<Sector> filterSectors(SectorStatus targetStatus, List<Sector> sectorList) {
```

Comment:
This should be in Provider (Role is making adapter List, kind of provider.)

Reply:
Thank you, modified.